src/vellum/workflows/context.py

```diff
@@ -0,0 +1,82 @@
+from contextlib import contextmanager
+from functools import wraps
+import inspect
+import threading
+from typing import Any, Callable, Dict, Iterator, TypeVar, Union, cast
+
+from vellum.workflows.events.types import ParentContext
+
+LoggingContext = Dict[str, Any]
+_CONTEXT_KEY = "_execution_context"
+
+local = threading.local()
+
+F = TypeVar('F', bound=Callable[..., Any])
+T = TypeVar('T')
+P = TypeVar('P', bound=ParentContext)
+
+
+def get_execution_context() -> LoggingContext:
+    """Retrieve the current execution context."""
+    return getattr(local, _CONTEXT_KEY, {})
```

Comment:
Once we make `ExecutionContext` a dataclass, I'm thinking it might be better to raise an exception if it's not set instead of returning something like `None`, so we don't have to do a null check every time we use it.